source/FreeRTOS_IP.c

```diff
@@ -1441,84 +1441,101 @@ BaseType_t xSendEventStructToIPTask( const IPStackEvent_t * pxEvent,
 eFrameProcessingResult_t eConsiderFrameForProcessing( const uint8_t * const pucEthernetBuffer )
 {
     eFrameProcessingResult_t eReturn = eProcessBuffer;
-    const EthernetHeader_t * pxEthernetHeader = NULL;
-    const NetworkEndPoint_t * pxEndPoint = NULL;
 
-    if( pucEthernetBuffer == NULL )
+    do
     {
-        eReturn = eReleaseBuffer;
-    }
-    else
-    {
-        /* Map the buffer onto Ethernet Header struct for easy access to fields. */
+        const EthernetHeader_t * pxEthernetHeader = NULL;
+        const NetworkEndPoint_t * pxEndPoint = NULL;
+
+        if( pucEthernetBuffer == NULL )
+        {
+            eReturn = eReleaseBuffer;
+            break;
+        }
 
+        /* Map the buffer onto Ethernet Header struct for easy access to fields. */
         /* MISRA Ref 11.3.1 [Misaligned access] */
         /* More details at: https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/MISRA.md#rule-113 */
         /* coverity[misra_c_2012_rule_11_3_violation] */
         pxEthernetHeader = ( ( const EthernetHeader_t * ) pucEthernetBuffer );
 
+        switch( pxEthernetHeader->usFrameType )
+        {
+            case ipARP_FRAME_TYPE:
+            case ipIPv4_FRAME_TYPE:
+                #if ipconfigIS_DISABLED( ipconfigUSE_IPv4 )
+                    eReturn = eReleaseBuffer;
+                #endif
+                break;
+
+            case ipIPv6_FRAME_TYPE:
+                #if ipconfigIS_DISABLED( ipconfigUSE_IPv6 )
+                    eReturn = eReleaseBuffer;
+                #endif
+                break;
+
+            default:
+                if( FreeRTOS_ntohs( pxEthernetHeader->usFrameType ) <= 0x600U )
+                {
+                    #if ipconfigIS_ENABLED( ipconfigFILTER_OUT_NON_ETHERNET_II_FRAMES )
+                        eReturn = eReleaseBuffer;
+                    #endif
+                }
+                else
+                {
+                    #if ipconfigIS_DISABLED( ipconfigPROCESS_CUSTOM_ETHERNET_FRAMES )
+                        eReturn = eReleaseBuffer;
+                    #endif
+                }
+                break;
+        }
+
         /* Examine the destination MAC from the Ethernet header to see if it matches
```

Comment:
Before this comment, I would expect the following conditional break:
~~~diff
        default:
            if( FreeRTOS_ntohs( pxEthernetHeader->usFrameType ) <= 0x600U )
            {
                #if ipconfigIS_ENABLED( ipconfigFILTER_OUT_NON_ETHERNET_II_FRAMES )
                    eReturn = eReleaseBuffer;
                #endif
            }
            else
            {
                #if ipconfigIS_DISABLED( ipconfigPROCESS_CUSTOM_ETHERNET_FRAMES )
                    eReturn = eReleaseBuffer;
                #endif
            }
            break;
    }    
+   if( eReturn == eReleaseBuffer )
+   {
+       /* No use to do more testing. */
+       break;
+   }
~~~

Reply:
Do you prefer that or using an if statement with breaks? I generally avoid places with multiple layers of breaks like switches in whiles.